File: backend/src/api/api_v1/routers/docker_ext.py
from fastapi import APIRouter, Depends, HTTPException
from src.datasources.docker import docker_datasource
from src.services.auth import auth_service
from src import models
from pydantic import BaseModel
from typing import List, Optional

router = APIRouter(prefix="/docker_ext", tags=["docker_ext"])


class ImageInfo(BaseModel):
    id: str
    repo_tag: str
    created: str


class NetworkInfo(BaseModel):
    id: str
    name: str
    driver: str
    # containers: dict  # необязательно отдавать все детали


class VolumeInfo(BaseModel):
    name: str
    created_at: str
    mountpoint: str
    scope: str


# ===== Images =====
@router.get("/images", response_model=List[ImageInfo])
async def list_images(current_user: models.User = Depends(auth_service.get_current_active_user)):
    return docker_datasource.list_images()


@router.delete("/images/{image_id}")
async def delete_image(image_id: str, current_user: models.User = Depends(auth_service.get_current_active_user)):
    try:
        docker_datasource.remove_image(image_id)
    except Exception as e:
        raise HTTPException(status_code=400, detail=f"Не удалось удалить образ: {e}")
    return {"status": "deleted"}


# ===== Networks =====
@router.get("/networks", response_model=List[NetworkInfo])
async def list_networks(current_user: models.User = Depends(auth_service.get_current_active_user)):
    return docker_datasource.list_networks()


@router.delete("/networks/{network_id}")
async def delete_network(network_id: str, current_user: models.User = Depends(auth_service.get_current_active_user)):
    try:
        docker_datasource.remove_network(network_id)
    except Exception as e:
        raise HTTPException(status_code=400, detail=f"Не удалось удалить сеть: {e}")
    return {"status": "deleted"}


@router.get("/networks/{network_id}/containers")
async def get_network_containers(
    network_id: str, current_user: models.User = Depends(auth_service.get_current_active_user)
):
    """
    Возвращает список контейнеров, подключённых к сети network_id.
    """
    try:
        lst = docker_datasource.get_containers_by_network(network_id)
    except Exception as e:
        raise HTTPException(status_code=400, detail=f"Ошибка при получении контейнеров сети: {e}")
    return {"containers": lst}


# ===== Volumes =====
@router.get("/volumes", response_model=List[VolumeInfo])
async def list_volumes(current_user: models.User = Depends(auth_service.get_current_active_user)):
    return docker_datasource.list_volumes()


@router.delete("/volumes/{volume_name}")
async def delete_volume(volume_name: str, current_user: models.User = Depends(auth_service.get_current_active_user)):
    try:
        docker_datasource.remove_volume(volume_name)
    except Exception as e:
        raise HTTPException(status_code=400, detail=f"Не удалось удалить том: {e}")
    return {"status": "deleted"}

File: frontend/src/lib/api/docker_ext.js
import { authFetch } from "$lib/api/auth";

/** ==== IMAGES ==== */
export async function getImages() {
  const res = await authFetch("/api/v1/docker_ext/images");
  if (!res.ok) throw new Error("Ошибка при получении списка образов");
  return await res.json();
}

export async function deleteImage(imageId) {
  const res = await authFetch(`/api/v1/docker_ext/images/${imageId}`, {
    method: "DELETE",
  });
  if (!res.ok) {
    const err = await res.json();
    throw new Error(err.detail || "Ошибка при удалении образа");
  }
  return true;
}

/** ==== NETWORKS ==== */
export async function getNetworks() {
  const res = await authFetch("/api/v1/docker_ext/networks");
  if (!res.ok) throw new Error("Ошибка при получении списка сетей");
  return await res.json();
}

export async function deleteNetwork(networkId) {
  const res = await authFetch(`/api/v1/docker_ext/networks/${networkId}`, {
    method: "DELETE",
  });
  if (!res.ok) {
    const err = await res.json();
    throw new Error(err.detail || "Ошибка при удалении сети");
  }
  return true;
}

export async function getNetworkContainers(networkId) {
  const res = await authFetch(
    `/api/v1/docker_ext/networks/${networkId}/containers`
  );
  if (!res.ok) throw new Error("Ошибка при получении контейнеров сети");
  return await res.json(); // { containers: [...] }
}

/** ==== VOLUMES ==== */
export async function getVolumes() {
  const res = await authFetch("/api/v1/docker_ext/volumes");
  if (!res.ok) throw new Error("Ошибка при получении списка томов");
  return await res.json();
}

export async function deleteVolume(volumeName) {
  const res = await authFetch(`/api/v1/docker_ext/volumes/${volumeName}`, {
    method: "DELETE",
  });
  if (!res.ok) {
    const err = await res.json();
    throw new Error(err.detail || "Ошибка при удалении тома");
  }
  return true;
}
